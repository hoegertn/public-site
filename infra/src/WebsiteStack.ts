import { Stack, StackProps } from 'aws-cdk-lib';
import { Certificate, CertificateValidation } from 'aws-cdk-lib/aws-certificatemanager';
import { Distribution, ViewerProtocolPolicy } from 'aws-cdk-lib/aws-cloudfront';
import { S3Origin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, HostedZone } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { BlockPublicAccess, Bucket, BucketEncryption } from 'aws-cdk-lib/aws-s3';
import { BucketDeployment, Source } from 'aws-cdk-lib/aws-s3-deployment';
import { Construct } from 'constructs';

export class WebsiteStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, props);

    const hostedZone = HostedZone.fromLookup(this, 'HostedZone', {
      domainName: 'openconstructfoundation.org',
    });

    const bucket = new Bucket(this, 'BucketForWebsite', {
      encryption: BucketEncryption.S3_MANAGED,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    });

    // Create CloudFront distribution
    const dist = new Distribution(this, 'Distribution', {
      defaultBehavior: {
        origin: new S3Origin(bucket),
        viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      },
      certificate: new Certificate(this, 'Certificate', {
        domainName: 'www.openconstructfoundation.org',
        validation: CertificateValidation.fromDns(hostedZone),
        subjectAlternativeNames: ['openconstructfoundation.org'],
      }),
      defaultRootObject: 'index.html',
      domainNames: ['www.openconstructfoundation.org', 'openconstructfoundation.org'],
    });

    new BucketDeployment(this, 'DeployWebsite', {
      sources: [Source.asset('../the-ocf-website/out')],
      destinationBucket: bucket,
      distribution: dist,
      distributionPaths: ['/*'],
    });


    new ARecord(this, 'WWWAlias', {
      zone: hostedZone,
      target: {
        aliasTarget: new CloudFrontTarget(dist),
      },
      recordName: 'www',
      deleteExisting: true,
    });
    new ARecord(this, 'RootAlias', {
      zone: hostedZone,
      target: {
        aliasTarget: new CloudFrontTarget(dist),
      },
      deleteExisting: true,
    });
  }

}
