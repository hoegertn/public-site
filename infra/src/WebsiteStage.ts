import { Stage, StageProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { WebsiteStack } from './WebsiteStack';

export class WebsiteStage extends Stage {
  constructor(scope: Construct, id: string, props: StageProps) {
    super(scope, id, props);
    new WebsiteStack(this, 'Website', {
      ...props,
    });
  }

}
