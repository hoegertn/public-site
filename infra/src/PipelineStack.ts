import { Stack, StackProps } from 'aws-cdk-lib';
import { CodePipeline, CodePipelineSource, ShellStep } from 'aws-cdk-lib/pipelines';
import { Construct } from 'constructs';
import { WebsiteStage } from './WebsiteStage';

export class PipelineStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    const pipeline = new CodePipeline(this, 'Pipeline', {
      synth: new ShellStep('Synth', {
        input: CodePipelineSource.connection('the-ocf/public-site', 'feat/new-content', {
          connectionArn: 'arn:aws:codestar-connections:us-east-1:935412892586:connection/d4dc7104-a60f-4f8f-b259-48b98cea4b22',
        }),
        primaryOutputDirectory: 'infra/cdk.out',
        commands: [
          'cd the-ocf-website',
          'npm install',
          'npm run build',
          'cd ../infra',
          'yarn',
          'yarn build',
          'npx cdk synth',
        ],
      }),
    });

    pipeline.addStage(new WebsiteStage(this, 'Prod', {
      ...props,
    }));
  }
}
