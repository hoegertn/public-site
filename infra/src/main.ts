import { App } from 'aws-cdk-lib';
import { PipelineStack } from './PipelineStack';


const app = new App();

new PipelineStack(app, 'OcfWebsitePipeline', {
  env: {
    account: '935412892586',
    region: 'us-east-1',
  },
});

app.synth();
