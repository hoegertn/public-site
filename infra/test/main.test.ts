import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { WebsiteStack } from '../src/WebsiteStack';

test('Snapshot', () => {
  const app = new App();
  const stack = new WebsiteStack(app, 'test', {
    env: {
      account: '123456789012',
      region: 'us-east-1',
    },
  });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
