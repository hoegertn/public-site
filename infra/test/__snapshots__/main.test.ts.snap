// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Mappings": {
    "AWSCloudFrontPartitionHostedZoneIdMap": {
      "aws": {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "BucketForWebsite85638C10": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:cr-owned:69b200d3",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketForWebsitePolicy0C2B365D": {
      "Properties": {
        "Bucket": {
          "Ref": "BucketForWebsite85638C10",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "DistributionOrigin1S3Origin5F5C0696",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "BucketForWebsite85638C10",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "e976a796f036a5efbf44b99e44cfb5a961df08d8dbf7cd37e60bf216fb982a00.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "DeployWebsiteAwsCliLayer17DBC421",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-123456789012-us-east-1",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-123456789012-us-east-1/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "BucketForWebsite85638C10",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "BucketForWebsite85638C10",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "cloudfront:GetInvalidation",
                "cloudfront:CreateInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomDeleteExistingRecordSetCustomResourceProviderHandlerAD00231E": {
      "DependsOn": [
        "CustomDeleteExistingRecordSetCustomResourceProviderRole03A7ED08",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "c07d810b6112c3008b1ee1bd76ea30ae40717d90e7b9922ff650d0cecf89e3bd.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomDeleteExistingRecordSetCustomResourceProviderRole03A7ED08",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomDeleteExistingRecordSetCustomResourceProviderRole03A7ED08": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "route53:GetChange",
                  "Effect": "Allow",
                  "Resource": "*",
                },
                {
                  "Action": "route53:ListResourceRecordSets",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":route53:::hostedzone/DUMMY",
                      ],
                    ],
                  },
                },
                {
                  "Action": "route53:ChangeResourceRecordSets",
                  "Condition": {
                    "ForAllValues:StringEquals": {
                      "route53:ChangeResourceRecordSetsActions": [
                        "DELETE",
                      ],
                      "route53:ChangeResourceRecordSetsRecordTypes": [
                        "A",
                      ],
                    },
                  },
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":route53:::hostedzone/DUMMY",
                      ],
                    ],
                  },
                },
                {
                  "Action": "route53:ListResourceRecordSets",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":route53:::hostedzone/DUMMY",
                      ],
                    ],
                  },
                },
                {
                  "Action": "route53:ChangeResourceRecordSets",
                  "Condition": {
                    "ForAllValues:StringEquals": {
                      "route53:ChangeResourceRecordSetsActions": [
                        "DELETE",
                      ],
                      "route53:ChangeResourceRecordSetsRecordTypes": [
                        "A",
                      ],
                    },
                  },
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":route53:::hostedzone/DUMMY",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DeployWebsiteAwsCliLayer17DBC421": {
      "Properties": {
        "Content": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "3322b7049fb0ed2b7cbb644a2ada8d1116ff80c32dca89e6ada846b5de26f961.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "DeployWebsiteCustomResourceD116527B": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketName": {
          "Ref": "BucketForWebsite85638C10",
        },
        "DistributionId": {
          "Ref": "Distribution830FAC52",
        },
        "DistributionPaths": [
          "/*",
        ],
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          "cdk-hnb659fds-assets-123456789012-us-east-1",
        ],
        "SourceObjectKeys": [
          "3f4431db433a0fc7e0142df9a81bd1e79b816f0da538e58b2ab33c493edabc24.zip",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "Distribution830FAC52": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "www.openconstructfoundation.org",
            "openconstructfoundation.org",
          ],
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "testDistributionOrigin1F4407531",
            "ViewerProtocolPolicy": "allow-all",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "BucketForWebsite85638C10",
                  "RegionalDomainName",
                ],
              },
              "Id": "testDistributionOrigin1F4407531",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "DistributionOrigin1S3Origin5F5C0696",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "ViewerCertificate": {
            "AcmCertificateArn": "arn:aws:acm:us-east-1:935412892586:certificate/16724c20-4370-4e56-8d7a-1c1d48022674",
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "DistributionOrigin1S3Origin5F5C0696": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Identity for testDistributionOrigin1F4407531",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "RootAliasCB7F041A": {
      "DependsOn": [
        "RootAliasDeleteExistingRecordSetCustomResourceBEF4D940",
      ],
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "Distribution830FAC52",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "openconstructfoundation.org.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "RootAliasDeleteExistingRecordSetCustomResourceBEF4D940": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "HostedZoneId": "DUMMY",
        "RecordName": "openconstructfoundation.org.",
        "RecordType": "A",
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomDeleteExistingRecordSetCustomResourceProviderHandlerAD00231E",
            "Arn",
          ],
        },
      },
      "Type": "Custom::DeleteExistingRecordSet",
      "UpdateReplacePolicy": "Delete",
    },
    "WWWAlias3083E442": {
      "DependsOn": [
        "WWWAliasDeleteExistingRecordSetCustomResource6F9A5702",
      ],
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "Distribution830FAC52",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "www.openconstructfoundation.org.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "WWWAliasDeleteExistingRecordSetCustomResource6F9A5702": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "HostedZoneId": "DUMMY",
        "RecordName": "www.openconstructfoundation.org.",
        "RecordType": "A",
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomDeleteExistingRecordSetCustomResourceProviderHandlerAD00231E",
            "Arn",
          ],
        },
      },
      "Type": "Custom::DeleteExistingRecordSet",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
